more_itertools/more.py

```diff
@@ -74,6 +74,32 @@
 _marker = object()
 
 
+def _decorator_factory(cls):
+    """Allow *cls* to wrap either an iterable or a function.
+    *cls* must be a class whose constructor takes an iterable, or a callable
+    function or method that returns an iterable object.
+
+    """
+    def decorator(iterable_or_callable):
+        try:
+            iter(iterable_or_callable)
+        except TypeError:
+            if not callable(iterable_or_callable):
+                raise
+        else:
+            return cls(iterable_or_callable)
+
+        @wraps(iterable_or_callable)
+        def callable_wrapper(*args, **kwargs):
+            return cls(iterable_or_callable(*args, **kwargs))
+
+        return callable_wrapper
+
+    decorator.__doc__ = cls.__doc__
```

Comment:
May be able to use @wraps for this as well

Reply:
Good call; `update_wrapper` seems to do the trick.